stdlib/universe/holt_winters.go

```diff
@@ -0,0 +1,463 @@
+		if err := hwt.craftKeyCols(builder, idx, v, newVs.Len()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getCleanData returns cleaned data (using the value and time column), and the first and last valid timestamps.
+// Below are the cleaning criteria.
+// Rows that have a null timestamp get discarded.
+// Rows that have a null value are considered invalid, but used by the algorithm.
+// HoltWinters supposes to work with evenly spaced values in time, so:
+//  - the Interval passed to the transformation is used to divide the data in time buckets;
+//  - if many values are in the same bucket, the first one is selected, the others are skipped;
+//  - if no value is present for a bucket, that is considered as an invalid value (treated like null values).
+// HoltWinters will only be provided with the values returned.
+// Timestamps can be deduced by summing interval to the first/last valid timestamp.
+func (hwt *holtWintersTransformation) getCleanData(tbl flux.Table, colIdx, timeIdx int) (*array.Float64, int64, int64, error) {
+	vs := array.NewFloat64Builder(fluxarrow.NewAllocator(hwt.alloc))
+	var start, stop int64
+	bucketEnd := int64(-1)
+	bucketFilled := false
+	roundTime := func(t int64) int64 {
+		interval := int64(hwt.interval)
+		remainder := t % interval
+		if remainder > interval/2 {
+			// round up
+			return (t/interval + 1) * interval
+		}
+		// round down
+		return (t / interval) * interval
+	}
+	nextBucket := func() {
+		bucketEnd += int64(hwt.interval)
+		bucketFilled = false
+	}
+	appendV := func(cr flux.ColReader, i int) {
+		switch typ := tbl.Cols()[colIdx].Type; typ {
+		case flux.TInt:
+			c := cr.Ints(colIdx)
+			if c.IsNull(i) {
+				vs.AppendNull()
+			} else {
+				vs.Append(float64(c.Value(i)))
+			}
+		case flux.TUInt:
+			c := cr.UInts(colIdx)
+			if c.IsNull(i) {
+				vs.AppendNull()
+			} else {
+				vs.Append(float64(c.Value(i)))
+			}
+		case flux.TFloat:
+			c := cr.Floats(colIdx)
+			if c.IsNull(i) {
+				vs.AppendNull()
+			} else {
+				vs.Append(float64(c.Value(i)))
+			}
+		default:
+			panic(fmt.Sprintf("cannot append non-numerical type %s", typ.String()))
+		}
+		bucketFilled = true
+	}
+	isNull := func(cr flux.ColReader, i int) bool {
+		switch typ := tbl.Cols()[colIdx].Type; typ {
+		case flux.TInt:
+			return cr.Ints(colIdx).IsNull(i)
+		case flux.TUInt:
+			return cr.UInts(colIdx).IsNull(i)
+		case flux.TFloat:
+			return cr.Floats(colIdx).IsNull(i)
+		default:
+			panic(fmt.Sprintf("cannot check non-numerical type %s", typ.String()))
+		}
+	}
+	isFirst := func() bool {
+		return bucketEnd == -1
+	}
+	if err := tbl.Do(func(cr flux.ColReader) error {
+		// we work row-wise
+		for i := 0; i < cr.Len(); i++ {
+			// drop values with invalid timestamp
+			if cts := cr.Times(timeIdx); cts.IsValid(i) {
+				// the first value must be valid, skip it if it isn't so
+				if isFirst() && isNull(cr, i) {
+					continue
+				}
+				trueT := cts.Value(i)
+				roundT := roundTime(trueT)
```

Comment:
is roundTime supposed to return the lower interval boundary?  it's always equal to either trueT  or trueT+interval.   

I think you can simplify this to  
1.  isFirst()  -->   record point and  bucketBegin = trueT; bucketEnd = bucketBegin + interval
2.  if  trueT < bucketEnd then continue
3.  else  record point and bucketBegin = bucketEnd ;  bucketEnd += interval

I can't prove it, but my gut feeling is that the 5 min. offset bug is hiding in this function.

Reply:
Got it, I'll further examine the behavior of the function.

I am supposing that this implementation is correct given that tests pass, and only the e2e test  that involves windowing fails.

The unit tests use the same dataset after `window -> first` run in InfluxQL.